Cryptocurrency Conversion Spice

claytonspinner opened the pull request:
#### What does your Instant Answer do?

Converts between cryptocurrencies, cryptocurrencies and traditional currencies, or lists popular conversions for a given cryptocurrency.

#### What problem does your Instant Answer solve (Why is it better than organic links)?

Solves for up-to-date conversion price between cryptocurrencies that may be listed on separate exchanges.

#### What is the data source for your Instant Answer? (Provide a link if possible)

cryptonator.com

#### Why did you choose this data source?

Supports a large number of cryptocurrencies (300+) and aggregates cryptocurrency spot prices from 52 exchanges. Has a public API without large restrictions. Averages prices from different exchanges and provides conversion for cryptocurrencies that may not be traded between each other (through BTC as the relative currency).

#### Are there any other alternative (better) data sources?

http://coinmarketcap.com/, though no public API is available without third party solutions such as http://coinmarketcap.northpole.ro/

Cryptocurrency exchanges themselves such as:
https://www.cryptsy.com/pages/publicapi
https://btc-e.com/api/documentation
https://www.bitstamp.net/api/
https://www.coins-e.com/exchange/api/documentation/
https://bter.com/api

though most of these only have a limited number of currencies or have a restricted API.

#### What are some example queries that trigger this Instant Answer?

"Convert ftc to ltc"
"25 ppc in btc"
"Lite coin"
"What is NXT in Bitcoins?"

#### Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)

Cryptocurrency traders. 

#### Is this Instant Answer connected to a DuckDuckHack Instant Answer idea?

Yes - https://duck.co/ideas/idea/4575/usd-eur-any-fiat-crypto-doge-ltc-digi-ftc-n

#### Which existing Instant Answers will this one supersede/overlap with?

DDG::Currency, DDG::Bitcoin. Traditional currencies (usd, aud, gbp) and "btc" are excluded from triggering DDG::Cryptocurrency when used by themselves. If they are used as part of a search with a cryptocurrency (for instance "50 usd to ltc") they will trigger DDG::Cryptocurrency.

#### Are you having any problems? Do you need our help with anything?

1. I can't get the links to work for different currencies on the item view of the spice. I'm not sure if this is just from being in the test environment?

2. None of the cryptocurrencies have flags. I don't mind adding the custom flags for the different currencies (with different sizes 92, 64, 32) but there are 300+ currencies and I don't know if you want all those pngs uploaded! Also, I added flags for LTC and FTC for testing, but I'm having trouble getting them to display. Again, not sure if this is an environment thing.

3. I copied the HTML and CSS directly from DDG::Currency and everything looks the same, but probably warrants double-checking.

#### What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?

Public API with no authentication. I reached out to cryptonator.com and they mentioned a limit on calls to the API when calling for a single cryptocurrency but I don't know what that limit is. They say on their site that "Should you wish to have individually customized API for your particular needs, feel free to contact us". If the limit becomes an issue we can probably reach out and request more bandwidth.

#### Where did you hear about DuckDuckHack? (For first time contributors)

From this post linked to on reddit: http://blog.ackx.net/duckduckhack-accepted.html

#### What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)

Conversion between cryptocurrencies:
![snapshot-conversion](https://cloud.githubusercontent.com/assets/7718370/6242809/bad6b6da-b703-11e4-98f5-8a5ee5bea067.PNG)

Single cryptocurrency search:
![snapshot-single](https://cloud.githubusercontent.com/assets/7718370/6242811/bee97f82-b703-11e4-836b-9f09efe85f1e.PNG)

Conversion between cryptocurrency (Mobile):
![screenshot_2015-02-18-00-22-22](https://cloud.githubusercontent.com/assets/7718370/6243063/5c8c763c-b708-11e4-894c-40a00b326467.png)

Single cryptocurrency search (Mobile):
![screenshot_2015-02-18-00-50-08](https://cloud.githubusercontent.com/assets/7718370/6243074/81e3a8b0-b708-11e4-9f14-b7d6c4950fb2.png)

### Checklist

Please place an 'X' where appropriate.

[X] Added metadata and attribution information
[X] Wrote test file and added to t/ directory
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [X] Google Chrome
        [X] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [] IE 10

    - Windows XP
        [] IE 7
        [] IE 8

    - Mac OSX
        [] Google Chrome
        [] Firefox
        [] Opera
        [] Safari

    - iOS (iPhone)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [X] Google Chrome
        [] Opera

https://duck.co/ia/view/cryptocurrency

MrChrisW commented:
@claytonspinner Thanks for your contribution, this will be reviewed shortly :+1:

jagtalon commented:
@claytonspinner thank you. Checking it out!

jagtalon commented on lib/DDG/Spice/Cryptocurrency.pm:
```diff
@@ -0,0 +1,174 @@
+triggers query_lc => qr/$currency_qr/;
+
+spice from => '([^/]+)/([^/]+)/([^/]*)';
+spice to => 'http://www.cryptonator.com/api/$1/$2?primary=$3';
```
Please change http to https--it redirects, and we don't allow the endpoints to redirect internally.

jagtalon commented:
@claytonspinner should copy in some of the assets in the currency spice as well for example https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/share/spice/currency/assets/32/xbt.png so that Bitcoin would work.

@abeyang maybe we should create our own assets for this, huh?

claytonspinner commented:
@jagtalon changed the endpoint to use https. copied the xbt.png from the Currency Spice. I renamed it to btc.png because the API used BTC to reference Bitcoin instead of XBT. Let me know if this is okay!

For all the other coins, the API supports over 300. I don't know if we should have flags for all of them or only a subset of the most popular ones.

CelestialWalrus commented:
@claytonspinner IMO you should use flags only for "top 10" (Ripple, Dogecoin, Litecoin, Namecoin, Peercoin, NXT, Darkcoin, Feathercoin, Blackcoin, Reddcoin).

http://cryptmarketcap.com/ has some nice logos.

moollaza commented:
@claytonspinner @CelestialWalrus if the assets are available in high enough resolution (e.g. for retina displays), I see no problem in using them. I just don't want to stretch smaller images because that obviously won't look very good.

@abeyang @thm any thoughts? If there weren't so many cryptocurrencies, I would say we could make our own assets, unless you feel like making a couple hundred icons @thm? :stuck_out_tongue_closed_eyes:

ScreapDK commented:
Hey, I just got an idea! Could we not just make the value of the `alt=""`-attribute be the crypto-currency "abbreviation"/"symbol" and style the heck out of it! xD

Somewhat like iOS 7.x does for phone contacts with no avatar! :D
I'll see if I can find a screenshot!

jagtalon commented:
@abeyang @moollaza @thm I'd say we just have one icon for all of them like https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/share/spice/currency/assets/96/xpt.png since most of them aren't that popular. Maybe just custom ones for litecoin and dogecoin specifically unless I'm wrong about their relative popularity

jagtalon commented:
Alright having a look at the code. @abeyang @moollaza in the mean time it's up on https://ddh9.duckduckgo.com/?q=convert+ltc+to+btc&ia=cryptocurrency

jagtalon commented:
@claytonspinner could you add this change to your code? Hovering / clicking on mobile shows this ugly underline. https://github.com/duckduckgo/zeroclickinfo-spice/pull/1572/files

Thanks!

MrChrisW commented:
@jagtalon @moollaza Regarding the icons, my preference would be to display a logo for the top 4/5 alt cryptocurrencies **only** based on market cap/popularity for now. 

The issue with using assets for alt-cryptocurrencies is maintainability, we'd need to update if a logo is changed, surprisingly this happens a lot. *Additional logos could be added by the community in the future if it's provided in the correct formats.*

I've generated some of the required assets (32,64,96), they can be [downloaded here](http://dev.mrchris.me/dl/cryptoassets.zip), I've only included Ripple as it's well known.

jagtalon commented:
@MrChrisW nice! Thanks for that. @abeyang check out the assets in https://github.com/duckduckgo/zeroclickinfo-spice/pull/1548#issuecomment-76436390 if they're okay!

jagtalon commented:
@moollaza @MrChrisW @abeyang I think the tab name should be "Currency" instead of "Cryptocurrency"--what do you think?

jagtalon commented:
@abeyang @chrismorast we've got an issue when it comes to really small numbers like:

![screen shot 2015-02-28 at 2 14 31 pm](https://cloud.githubusercontent.com/assets/81969/6427650/a7d54b88-bf54-11e4-9f0b-2add77207c6a.png)

And it's exactly the thing that's brought up in https://github.com/duckduckgo/zeroclickinfo-spice/issues/1465. I think we should extend it to more than two decimal places as mentioned by @MrChrisW but maybe **only if the first two are just zeroes**. What do you think?

claytonspinner commented:
@jagtalon @MrChrisW Thanks Chris! I added these to the assets. Also, any coin without a flag will use [default.png](https://github.com/claytonspinner/zeroclickinfo-spice/blob/cryptocurrency-conversion/share/spice/cryptocurrency/assets/92/default.png)

@jagtalon I added the formatting to decimals for small numbers. If the value is smaller than the decimal limit, it will print to the first significant digit that isn't a zero with rounding. So '0.0036 USD' will display as '0.004 USD'. Traditional currencies display 2 decimal places, cryptocurrencies up to 8. In addition, the two conversion rates displayed on the right will display up to 8 decimal places. Here's an example:
![snapshot-decimal](https://cloud.githubusercontent.com/assets/7718370/6428987/ba6c8802-bf85-11e4-917f-b4a94bfd0ffa.PNG)
I can change any of the values for formatting, just let me know!

abeyang commented:
Hi @claytonspinner , the design team (namely, @chrismorast) created our own assets for the more popular ones, using the official DDG color schemes. I've included them here (both 1x and 2x):
http://cl.ly/2e3v2E2X2L1R

Here's what they'll look like:
![screen shot 2015-03-18 at 2 14 54 pm](https://cloud.githubusercontent.com/assets/794084/6743294/448aba80-ce55-11e4-9077-85c6ce7ce2f8.png)

MrChrisW commented:
@abeyang @chrismorast  Fantastic looking flat style icons! 
Do you have any comments regarding displaying [8 decimal places](https://github.com/duckduckgo/zeroclickinfo-spice/pull/1548#issuecomment-76541115) ?

chrismorast commented:
@MrChrisW and @claytonspinner , :+1: on the decimal solution - as long is it wouldn't get clipped in any situation that the content might expand outside of the usable space.

claytonspinner commented:
@abeyang @chrismorast Thanks guys, these look awesome! I've added them in. I also changed the default decimal length for cryptocurrencies to 4 decimal spaces, I think it looks a little better than 8.

jagtalon commented:
@abeyang @chrismorast that looks really cool! @claytonspinner sorry for the delay--checking it out!

jagtalon commented:
@claytonspinner looks like the dir should be called `96` instead of `92`. We're not getting the images on high-def screens as a result:

![screen shot 2015-04-08 at 4 06 56 am](https://cloud.githubusercontent.com/assets/81969/7041139/be83affa-dda4-11e4-9902-0f01942f1e82.png)

jagtalon commented:
Looks like default is different in 96 and 32

![screen shot 2015-04-08 at 4 17 16 am](https://cloud.githubusercontent.com/assets/81969/7041297/34acaf64-dda6-11e4-86ae-315fc789f8e7.png)
![screen shot 2015-04-08 at 4 17 19 am](https://cloud.githubusercontent.com/assets/81969/7041298/34ae7916-dda6-11e4-90f1-5eda6853922a.png)

jagtalon commented on share/spice/cryptocurrency/cryptocurrency.js:
```diff
@@ -0,0 +1,250 @@
+        "drk": true,
+        "ltc": true,
+        "xrp": true
+    }
```
Should you add the other cryptocurrencies in here now that you have more? Queries such as "convert ltc to qrk" will still show qrk with a default flag.

jagtalon replied:
Currently looks like:

![screen shot 2015-04-08 at 4 33 26 am](https://cloud.githubusercontent.com/assets/81969/7041576/7b993daa-dda8-11e4-8425-1508f3630782.png)

But should be:

![screen shot 2015-04-08 at 4 33 44 am](https://cloud.githubusercontent.com/assets/81969/7041579/803f5ab0-dda8-11e4-94b6-83e067f0ec71.png)

claytonspinner commented:
@jagtalon Whoops! I changed /92 to /96 and changed /96/generic.png to be the same as /32/generic.png. Also, the design team provided flags for 1x and 2x but not 3x. I resized the 2x flags and used them for 3x. Let me know if this is OK. Thanks!

jagtalon commented:
Checking it!

jagtalon commented:
@claytonspinner 

![screen shot 2015-04-14 at 4 40 48 pm](https://cloud.githubusercontent.com/assets/81969/7146948/34f622c4-e2c5-11e4-959c-756726fe95f2.png)

- The borders are a bit off. Check out https://github.com/duckduckgo/zeroclickinfo-spice/commit/ab2df2d3af42f5190801fb80680066d863b465e8. I think that will fix it.

- The bottom text is missing in the first tile.

- @abeyang what do you think of https://github.com/duckduckgo/zeroclickinfo-spice/pull/1548#issuecomment-92100289? It's up on https://ddh9.duckduckgo.com/?q=50+ltc+to+usd&ia=cryptocurrency btw.

claytonspinner commented:
@jagtalon Fixed the border. Bottom text for the first tile should be displaying now as well.

abeyang commented:
Looks good! However, clicking on the "switch currency" icon doesn't work for some of them...

![pasted_image_4_24_15__10_53_am](https://cloud.githubusercontent.com/assets/794084/7325028/4b9ae610-ea70-11e4-918e-13f223487c24.png)

chrismorast commented:
What determines which cryptocurrency conversions show up on this query? We should try to make sure we create icons for any of the top currencies (which we did but looks like we missed most of the ones displaying here). The placeholder icon (shown as a blank gold circle) should really only show up for the more obscure cryptocurrencies - probably more like a specific query for that conversion.

ScreapDK commented:
@abeyang First... (I don't think we've "met" before)... so.. Hello! Nice to meet you! :D What I actually wanted to was to ask where that screenshot is from. It's just that I'm not able to replicate/activate the IA that is active in the aforemenitioned screenshot.

I'm just wondering if there exists a more or less "public development/test/???"-server, where WIP Instant Answers can be tes? Just that I'm really excited for this particular IA to go live, that I almost cait wait! ^_^ :D (cc: @zekiel)

MrChrisW commented:
Hey @ScreapDK I'll anwser as I think @abeyang is probably sleeping :laughing: 

Take a look at the ["Development Pipeline"](https://duck.co/ia/pipeline/dev) on duck.co 

The Cryptocurrency Spice is listed and provides a link to the DDH9 server where it's deployed. 
+ https://duck.co/ia/view/cryptocurrency
+ https://ddh9.duckduckgo.com/?q=50+ltc&ia=cryptocurrency

It may not be the most up to date version. The other option is [pulling pull requests locally](https://help.github.com/articles/checking-out-pull-requests-locally/) and running `duckpan server <IA name>` :+1:

claytonspinner commented:
@abeyang While looking this up I discovered that cryptonator.com has recently removed a lot of the smaller one-off exchanges from its listing. Here's an archived version of the listing for BLA:

https://web.archive.org/web/20150403190326/https://www.cryptonator.com/rates#BLA

The broken links seem to correspond with the currencies that had been removed (except BTC, which is still listed as a trade currency for BLA). When I search "1 BLA"  now I only get a few results:
![bla_example](https://cloud.githubusercontent.com/assets/7718370/7442626/91520f86-f0e6-11e4-9da4-167fad96910a.PNG)
I still have to dig deeper to see if I can understand exactly what happens when exchanges are added or removed from the underlying API, but I think the links breaking are related to that.

@chrismorast the currencies that are returned for a search are the ones that the search currency can be directly exchanged to and from on the supported exchanges. A lot of the less popular currencies are only exchanged to a from BTC, so BTC is the only box that comes back (this may not be the best user experience). I did a count of the most popular exchange currencies and these are the numbers.

    {
        "BTC": 563,
        "USD": 33,
        "XRP": 98,
        "LTC": 138,
        "CNY": 49,
        "DOGE": 115,
        "XMR": 10,
        "HTML5": 68,
        "AUD": 1,
        "EUR": 3,
        "GBP": 3,
        "HKD": 1,
        "JPY": 1,
        "NZD": 1,
        "PLN": 3,
        "RUR": 5,
        "SGD": 1,
        "NXT": 3,
        "USDT": 4,
        "XPM": 1,
        "PPC": 1
    }

chrismorast commented:
Thanks @claytonspinner ! I may just try to improve the placeholder icon. I'll play around with it a bit and get something over to you later today.

moollaza commented:
Hey @claytonspinner! Sorry for the delay here -- I haven't forgotten about this PR!

I have some great news, we found an awesome Icon font that had great image assets for Crypto Currency icons. Our design team has improved them and would like us to use them for this IA. The assets can be found here: http://cl.ly/083p3a2M0L3X

We should use the SVG's unless SVG usn't supported. You can check for SVG support like so: https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/share/spice/sun_rise_set/sun_rise_set.js#L37, using the global Modernizr object

claytonspinner commented:
Thanks @moollaza! I'll add these in. I see just 32px PNG and not 64 and 96. Should I scale up the 32px PNGs to the larger sizes and include them?

moollaza commented:
>I see just 32px PNG and not 64 and 96. Should I scale up the 32px PNGs to the larger sizes and include them?

@claytonspinner I don't believe we'll need to use larger PNG's? PNG's will likely only be applied on IE otherwise SVG should be OK and can obviously scale if needed.

claytonspinner commented:
@moollaza Okay, thanks. It should be all set then.

moollaza commented:
@claytonspinner **so** sorry! Somehow your reply slipped by me. This look great! Thank a ton for the awesome work, we really appreciate it!

Again, sorry for the delay -- I'm trying to make sure these PR's move along a lot smoother :smile: 

- https://www.browserstack.com/screenshots/3dff5b2b37c470c6667761c5a31dcc1b11dd7e9d
- https://www.browserstack.com/screenshots/becc1afb9c927ff138c7cf9d2cce316e8373c99c

moollaza commented on share/spice/cryptocurrency/cryptocurrency.js:
```diff
@@ -0,0 +1,292 @@
+                rate = "1 " + target + " = " + formatNumber((1 / price), base, 8) + " " + base,
+                inverseRate = "1 " + base + " = " + formatNumber(price, target, 8) + " " + target,
+                // Format Time and Date
+                timestamp = (new Date(api_result.timestamp*1000)).toISOString(),
```
Just noticed that this is breaking IE8. It looks like `toISOString` is only supported in IE9+. I'll try and fix with Moment.js.

moollaza commented:
@claytonspinner congrats! It's live: https://duckduckgo.com/?q=50+ltc&ia=cryptocurrency

Sorry for the long delay! 

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's, tackle open issues or even submit more Instant Answers!

claytonspinner commented:
@moollaza This is awesome! Thanks again for all the support while making this. I'm looking forward to contributing in the future!

tmlee commented:
Hey guys, just wanted to share with you guys our recently launched public API at https://www.coingecko.com/api/docs/v3 up for consideration as an option.

By calling https://api.coingecko.com/api/v3/coins/bitcoin, you get a whole bunch of data back from our service. The one attribute that would be of interest for this would be the `current_price`, which returns price across multiple different national currencies.

We have other data supported by other endpoints as well, happy to learn and understand your need and to see how we can help accommodate.
